libraries/c_sdk/aws/defender/src/aws_iot_defender_collector.c

```diff
@@ -405,6 +457,212 @@ static void _serializeTcpConnections( void * param1,
+                                                               IotSerializer_ScalarSignedInt( pxTaskStatusArray[ i ].eCurrentState ) );
+            assertNoError( serializerError );
+
+            serializerError = _defenderEncoder.appendKeyValue( &taskEntryMap,
+                                                               KERNEL_TASK_PRIORITY,
+                                                               IotSerializer_ScalarSignedInt( pxTaskStatusArray[ i ].uxCurrentPriority ) );
+            assertNoError( serializerError );
+
+            serializerError = _defenderEncoder.appendKeyValue( &taskEntryMap,
+                                                               KERNEL_TASK_ABS_CYCLES,
+                                                               IotSerializer_ScalarSignedInt( pxTaskStatusArray[ i ].ulRunTimeCounter ) );
+            assertNoError( serializerError );
+
+            /* if idle, save the idle time counter. */
+            if( TASK_NAME_EQUAL == _taskNameFilterIdle( pxTaskStatusArray[ i ].pcTaskName ) )
+            {
+                ulIdleTime = pxTaskStatusArray[ i ].ulRunTimeCounter;
+            }
+
+            serializerError = _defenderEncoder.appendKeyValue( &taskEntryMap,
+                                                               KERNEL_TASK_PERCENTAGE,
+                                                               IotSerializer_ScalarSignedInt( pxTaskStatusArray[ i ].ulRunTimeCounter / ( ulTotalTime / 100 ) ) );
+
+
+            assertNoError( serializerError );
+
+            serializerError = _defenderEncoder.appendKeyValue( &taskEntryMap,
+                                                               KERNEL_STACK_HIGH_WATERMARK,
+                                                               IotSerializer_ScalarSignedInt( pxTaskStatusArray[ i ].usStackHighWaterMark ) );
+            assertNoError( serializerError );
+
+            serializerError = _defenderEncoder.closeContainer( &taskDetailsArray, &taskEntryMap );
+            assertNoError( serializerError );
+        }
+
+        /* Close array container */
+        serializerError = _defenderEncoder.closeContainer( &mcuUptimeMap, &taskDetailsArray );
+        assertNoError( serializerError );
+
+        /* mcu_uptime, mcu_utilization. */
+        serializerError = _defenderEncoder.appendKeyValue( &mcuUptimeMap,
+                                                           KERNEL_MCU_UPTIME,
+                                                           IotSerializer_ScalarSignedInt( ulTotalTime ) );
+        assertNoError( serializerError );
+
+        /* Calculate for MCU utilization, given now we know how many cycles are spent in IDLE. */
+        serializerError = _defenderEncoder.appendKeyValue( &mcuUptimeMap,
+                                                           KERNEL_MCU_UTILIZATION,
+                                                           IotSerializer_ScalarSignedInt( ( ulTotalTime - ulIdleTime ) / ( ulTotalTime / 100 ) ) );
```

Comment:
Prefer not, since potential overflow. Followed the same logic in kernel. https://github.com/aws/amazon-freertos/blob/master/freertos_kernel/tasks.c#L4463

You do have a point that "how early can this function be executed", can it be before 100 cycles of timer? Likely not, since the requirement for perf counter frequency is at least 10 times faster than SysTick... There will be a check for timer frequency configuration to make sure the frequency is within expectation, so that divide by zero does not happen.

Reply:
ok, the kernel code that you pointed to also handles "zero ulTotalTime". We can do:

```
( ulTotalTime - ulIdleTime )  / ulTotalTime * 100
```

Reply:
hmmm it will not work either. Dividend is always larger than divisor. And as a result of integer rule, the result is always zero.

Reply:
Ah, my bad.  If the overflow is concerned more than divide-by-zero, I agree with your original code.